src/Command/Backup/BackupDatabaseCommand.php

```diff
@@ -0,0 +1,161 @@
+use RuntimeException;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Process\Process;
+use Throwable;
+use UnexpectedValueException;
+
+class BackupDatabaseCommand extends Command
+{
+    protected static $defaultName = 'andreas-a:backup:database';
+
+    /**
+     * @var Filesystem
+     */
+    private $filesystem;
+
+    /**
+     * @var string
+     */
+    private $targetDirectory;
+
+    /**
+     * @var BackupDatabaseHandlerChain
+     */
+    private $handler;
+
+    public function __construct(
+        BackupDatabaseHandlerChain $handler,
+        Filesystem $filesystem,
+        string $targetDirectory,
+        string $name = null
+    ) {
+        parent::__construct($name);
+
+        $this->filesystem = $filesystem;
+        $this->handler = $handler;
+        $this->targetDirectory = rtrim($targetDirectory, '/');
+    }
+
+    protected function configure()
+    {
+        $this->setDescription('Creates a database backup.');
+    }
+
+    /**
+     * @throws Throwable
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $io = new SymfonyStyle($input, $output);
+        $io->section('Starting database backup.');
+
+        $currentDateTime = (new DateTime())->format(DateTimeInterface::RFC3339_EXTENDED);
+
+        $backupPath = sprintf('%1$s/db_%2$s.sql.bz2', $this->targetDirectory, $currentDateTime);
+
+        $this->filesystem->mkdir($this->targetDirectory);
+
+        $namedPipe = $this->createNamedPipe();
+        $compressProcess = null;
+
+        try {
+            $compressProcess = $this->createCompressProcess($namedPipe, $backupPath);
+            $compressProcess->start();
+
+            $dumpProcess = $this->handler->createDumpProcess($namedPipe);
+            $dumpProcess->run();
+
+            if (false === $dumpProcess->isSuccessful()) {
+                throw new RuntimeException($dumpProcess->getErrorOutput());
+            }
+
+            $compressProcess->wait();
+        } catch (Throwable $throwable) {
+            if (null !== $compressProcess && $compressProcess->isRunning()) {
+                $compressProcess->stop(0);
+            }
+
+            throw $throwable;
+        } finally {
+            // Remove potentially created named pipe.
+            $this->filesystem->remove($namedPipe);
+        }
+
+        $errorOutput = $dumpProcess->getErrorOutput();
+        $io->getErrorStyle()->write($errorOutput);
+
+        $isFailure =
+            false === $dumpProcess->isSuccessful() ||
+            false === $compressProcess->isSuccessful() ||
+            // Some errors like missing PROCESS privilege result in success exit code but with corresponding error.
+            0 === strpos($errorOutput, 'mysqldump: Error: ');
```

Comment:
This is hander specific and should reside in the handler.

Reply:
already moved it 😄